utils/core/src/serde/byte_reader.rs

```diff
@@ -92,6 +92,26 @@ pub trait ByteReader {
         Ok(u64::from_le_bytes(bytes))
     }
 
+    /// Returns a usize value read from `self` in ULEB128 format.
+    ///
+    /// # Errors
+    /// Returns a [DeserializationError] if a usize value could not be read from `self`.
+    fn read_usize(&mut self) -> Result<usize, DeserializationError> {
+        let mut result = usize::default();
+        let mut offset = 0;
+        loop {
+            let octet = self.read_u8()?;
+            // get the value from the octet and accumulate it in the result
+            result += (octet & OCTET_BIT_MASK) as usize * 128usize.pow(offset);
+            // if the continuation bit equals 0, finish reading
+            if octet >> OCTET_VALUE_SIZE == 0 {
+                break;
+            }
+            offset += 1;
+        }
```

Comment:
We should probably check if the value we read is within acceptable bounds for `usize` on a given platform - and if not, return an error.